### PR TITLE
drivers: imx_i2c: add support for imx6ull

### PR DESCRIPTION
**Pull request by ldts**
Support native I2C access on imx6ull (this SoC has an additional
register to configure the daisy chain in the iomuxc).

Tested on imx6ull-evk 14x14 with the bus at 400Kbps.

A patch has been sent upstream to address the current configuration in U-Boot.
https://lists.denx.de/pipermail/u-boot/2020-October/430482.html

### CODE REVIEW

**Comment by ldts** [comment]
@jenswi-linaro any chance this can be merged in time for 3.11?

File: core/drivers/imx_i2c.c
Comment (jforissier) [inline core/drivers/imx_i2c.c:140]: Use of `if (IS_ENABLED(CFG_MX8MM)) { ... }` from `<config.h>` is preferred.

**Review by jforissier — COMMENTED** [review]
With comment below addressed:
`Acked-by: Jerome Forissier <jerome@forissier.org>`

File: core/drivers/imx_i2c.c
Comment (ldts) [reply core/drivers/imx_i2c.c:140]: ok will fix.
I also need to review some SPL settings which might have been hiding an issue on this PR.

Comment (ldts) [reply core/drivers/imx_i2c.c:152]: about this, unfortunately we cant use it in this instance as the imx6ull uses some definitions that are not defined for mx8mm...ie BM_CCM_CCGR2_I2C1_SERIAL

**Comment by ldts** [comment]
@jforissier I pushed an update: imx6ull uses an aditional register to configure the daisy chain in the iomuxc; while testing I noticed that this register was being configured by SPL and never reprogrammed in the i2c driver meaning that i2c wouldnt be accessible without those SPL/bootloader versions.
it is all good now.

File: core/drivers/imx_i2c.c
Comment (jforissier) [reply core/drivers/imx_i2c.c:152]: Ah, yes. #ifdef is OK then.

**Comment by jforissier** [comment]
LGTM

File: core/arch/arm/plat-imx/registers/imx6.h
Comment (ldts) [inline core/arch/arm/plat-imx/registers/imx6.h:144]: if the BSP configured the peripheral clock to run of the 24MHz oscillator, the definition above is unnecessary.

Comment (ldts) [reply core/arch/arm/plat-imx/registers/imx6.h:145]: I added a comment to the imx6.h file; since the upstream uboot uses the 66MHz clock instead of the 24MHz oscillator.

Comment (jforissier) [reply core/arch/arm/plat-imx/registers/imx6.h:145]: Can't you tell the clock from DT?

Comment (ldts) [reply core/arch/arm/plat-imx/registers/imx6.h:145]: yes...I need to add proper dt support; I send a patch upstream (uboot) to default to the 24MHz oscillator...let me remove this support for 66MHz so we can get it merge and then will work out the dt configuration as we add support for more boards. hope that is acceptable

Comment (jforissier) [reply core/arch/arm/plat-imx/registers/imx6.h]: Sure, DT support can be added later. In the mean time I would recommend using a config variable in `plat-imx/conf.mk`, for instance `CFG_IMX_i2C_CLK_RATE ?= 24000000` (or whichever value is more convenient as a default...).

Comment (ldts) [reply core/arch/arm/plat-imx/registers/imx6.h]: ok

File: core/drivers/imx_i2c.c
Comment (jforissier) [inline core/drivers/imx_i2c.c:17]: You could also use the CFG_ directly

File: core/arch/arm/plat-imx/conf.mk
Comment (jforissier) [inline core/arch/arm/plat-imx/conf.mk:191]: Why not just add:
```
CFG_IMX_I2C_CLK_RATE ?= 24000000
```
...or perhaps:
```
CFG_IMX_I2C ?= n
CFG_IMX_I2C_CLK_RATE ?= 24000000
```
...near the end of the file, for all boards? It doesn't matter much if variables are defined but not used.
I let other iMX guys comment based on their preference ;-)

**Review by jforissier — COMMENTED** [review]
Should be one single commit IMO.

File: core/arch/arm/plat-imx/conf.mk
Comment (ldts) [reply core/arch/arm/plat-imx/conf.mk:191]: yes I like your suggestion better...shall I wait for comments or just go with your proposal?

**Comment by ldts** [comment]
anything I can do to help with it being merged  on time for the 3.11 release?

**Comment by jforissier** [comment]
> anything I can do to help with it being merged on time for the 3.11 release?

Paying more attention to review comments would certainly help.

**Review by etienne-lms — COMMENTED** [review]
`Acked-by: Etienne Carriere <etienne.carriere@linaro.org>`

**Comment by ldts** [comment]
> > anything I can do to help with it being merged on time for the 3.11 release?
> 
> Paying more attention to review comments would certainly help.

Yes, and that goes both ways

**Comment by jforissier** [comment]
@ldts if I'm not mistaken, none of the three comments I made here: https://github.com/OP-TEE/optee_os/pull/4152#pullrequestreview-508379818 were addressed properly.

File: core/drivers/imx_i2c.c
Comment (clementfaure) [inline core/drivers/imx_i2c.c:156]: IMO, adding another macro would be better for consistency with the lines above.
```#define CCGR2_I2C_SERIAL(_x) BM_CCM_CCGR2_I2C##_x_SERIAL```
```val = CCGR2_I2C_SERIAL(bid)```

File: core/arch/arm/plat-imx/registers/imx6.h
Comment (ldts) [reply core/arch/arm/plat-imx/registers/imx6.h:144]: sure, that can be done.
why do you say that CFG_IMX_I2C_CLK_RATE is the same for all supported platforms? where did you get that information from? the peripheral clock rate is not always configured  with the oscillator clock (certainly not in mainline u-boot)

File: core/arch/arm/plat-imx/registers/imx6.h
Comment (ldts) [reply core/arch/arm/plat-imx/registers/imx6.h:148]: shouldnt that better be done once someone merges an iomuxc driver?

File: core/drivers/imx_i2c.c
Comment (ldts) [reply core/drivers/imx_i2c.c:156]: ok. I think we missed this 3.11 window. I'll post something for the next release. closing it now.

File: core/arch/arm/plat-imx/registers/imx6.h
Comment (clementfaure) [reply core/arch/arm/plat-imx/registers/imx6.h:148]: The IOMUX controller has a bunch of registers that are meant to be configured by the other peripheral drivers. No need to wait for an iomux driver to organize things a bit and avoid messy header file.
Just like ```imxX-crm.h``` header file, we don't have a specific driver for clocks, but yet, we got a separate header file.

Comment (ldts) [reply core/arch/arm/plat-imx/registers/imx6.h:148]: @clementfaure so this file (imx6-iomux.h),  shall I just pick it from code aurora?

**Comment by ldts** [comment]
@clementfaure added support for DT and pulled imx6-iomux.h from Codeaurora.  have a good weekend.

File: core/drivers/imx_i2c.c
Comment (clementfaure) [inline core/drivers/imx_i2c.c:331]: I think  ```TEE_ERROR_BAD_STATE``` is more adequate since this error is not related to parameters.

File: core/drivers/imx_i2c.c
Comment (clementfaure) [inline core/drivers/imx_i2c.c:372]: Same thing

File: core/drivers/imx_i2c.c
Comment (clementfaure) [inline core/drivers/imx_i2c.c:396]: Same thing

**Review by clementfaure — CHANGES_REQUESTED** [review]

File: core/arch/arm/plat-imx/registers/imx6-crm.h
Comment (ldts) [reply core/arch/arm/plat-imx/registers/imx6-crm.h:861]: why? dont they belong with the clock?

File: core/arch/arm/plat-imx/registers/imx6-iomux.h
Comment (ldts) [reply core/arch/arm/plat-imx/registers/imx6-iomux.h:95]: why? dont they belong with the iomux?

File: core/arch/arm/plat-imx/registers/imx6-iomux.h
Comment (ldts) [reply core/arch/arm/plat-imx/registers/imx6-iomux.h:1]: which ones? only the ones we use? I dont think we use any

File: core/drivers/imx_i2c.c
Comment (ldts) [reply core/drivers/imx_i2c.c:331]: I'd rather keep bad parameters: the way I see it, the user requested a read transaction for a bus that it didnt initialize

File: core/drivers/imx_i2c.c
Comment (ldts) [reply core/drivers/imx_i2c.c:372]: same as before

File: core/drivers/imx_i2c.c
Comment (ldts) [reply core/drivers/imx_i2c.c:396]: same as before

File: core/arch/arm/plat-imx/registers/imx6-iomux.h
Comment (ldts) [reply core/arch/arm/plat-imx/registers/imx6-iomux.h:93]: why now? for consistency I'd also have mot modify the ones in imx8m.h?

File: core/arch/arm/plat-imx/registers/imx6.h
Comment (ldts) [reply core/arch/arm/plat-imx/registers/imx6.h:121]: yes that is the idea: to be able to work around bootloaders configuring the peripheral clock at a different rate (nowhere says that 24MHz is mandatory).
so where do you want this clock config?

**Comment by ldts** [comment]
@clementfaure to avoid the back and forth I pushed an update to the PR. let me know if this is what you are after.

File: core/arch/arm/plat-imx/registers/imx6.h
Comment (clementfaure) [reply core/arch/arm/plat-imx/registers/imx6.h:121]: Well, I would leave the default clock speed as a ```#define ...``` like it was before.
I would like to avoid defining the clock speed of a peripheral in ```conf.mk``` becoming the norm ... I don't know where we should move this.
Is the default clock speed working with se050?
If you need to modify the clock speed, is it possible to keep the modification in your own tree?

File: core/arch/arm/plat-imx/registers/imx6-iomux.h
Comment (clementfaure) [reply core/arch/arm/plat-imx/registers/imx6-iomux.h:1]: Alright, I guess it's time to remove unused definitions.
Could you do it?

File: core/include/drivers/imx_i2c.h
Comment (clementfaure) [inline core/include/drivers/imx_i2c.h:17]: I think you misread my comment https://github.com/OP-TEE/optee_os/pull/4152#discussion_r511817381 
I only asked if you could move ```I2C_``` definitions, not ```IOMUXC```

 ```IOMUXC``` can definitly go to ```imx8m.h``` header file

**Review by clementfaure — CHANGES_REQUESTED** [review]

File: core/arch/arm/plat-imx/registers/imx6.h
Comment (jforissier) [reply core/arch/arm/plat-imx/registers/imx6.h:121]: > I would like to avoid defining the clock speed of a peripheral in `conf.mk` becoming the norm ...

What's wrong with that?

Comment (clementfaure) [reply core/arch/arm/plat-imx/registers/imx6.h:121]: There are already a lot of stuff in ```conf.mk```
If we go in this direction, we could define a default clock speed for each peripheral and/or each platform ... adding even more things to this file. That's something I would like to avoid if possible.

Comment (jforissier) [reply core/arch/arm/plat-imx/registers/imx6.h:121]: > There are already a lot of stuff in `conf.mk`

Well, if it is not in `conf.mk` it will be somewhere else ;-) If things get too big we can always create separate `.mk` files and include them (one `.mk` file for a specific device/driver for example).

> If we go in this direction, we could define a default clock speed for each peripheral and/or each platform 

;-) depends on the device I suppose.
The main thing I would like to emphasise is: in general, fixed parameters are best defined in header files (with `#define`), but configurable parameters are much easier in `.mk` files because with one line `CFG_SOMETHING ?= value` you can define a default value, let the user override it easily if needed, and make it available to makefiles and C code.

That being said I have no strong opinion, and @clementfaure I will happily leave you the final word.

File: core/include/drivers/imx_i2c.h
Comment (ldts) [reply core/include/drivers/imx_i2c.h:17]: It is a bit confusing because you are asking for offsets that are defined as IOMUXC_I2C1_SCL_CFG_OFF in imxX.h but offsets defined using macros (ie for I2C2 and I2C3) in a different file which is counter intuitive.

And we are also deviating from what was asked when support for imx8m was merged.

if it was down to me I'd still write:
1) offsets in imx8m.h 
2) hardcoded target values for clock and iomux (since we dont have those drivers)
 - either in the corresponding iomux.h and crm.h 
 - or even better just in the imx_i2c.c driver itself

anyway, I'll try again

Comment (ldts) [reply core/include/drivers/imx_i2c.h:17]: @clementfaure I am going to close this PR and open a new one from which I hope to get some clarity on the review comments (some of the things you are asking don't quite make much sense to me).

**1) moving I2C_CFG_ to imx_i2c.h** 

why? it does not belong in imx_i2c.h...perhaps in imx_i2c.c since some I2C_ are utility macros; however not all of them are: others are hardcoded clock and iomux calculated values...so where do you expect these later values to land? and why cant we do now as it was done for imx8mm.?

2)  importing an imx6-iomux.h 
why? the i2c driver will not use any values from that file so why do we need it done in this PR.

3) dropping support for boot-loaders not setting the peripheral clock to 24MHz
why? can you provide a valid reason for that.

so please bear me, let me send the PR again, and then we can discuss and hopefully align.

**Comment by ldts** [comment]
#4186

File: core/include/drivers/imx_i2c.h
Comment (clementfaure) [reply core/include/drivers/imx_i2c.h:17]: @ldts 
Alright, no problem, I will have a look at the new PR.
I suggest we find a solution in the comment section that suit both of us first, and then push changes. By doing so, I believe we would avoid confusion and new pull-requests. What do you think?

**1/ Moving ```I2C_CFG_``` to ```imx_i2c.h```**
I believe ```I2C_CFG_``` macros should belong to ```imx_i2c.h``` as the name of the macro suggest, same thing for ```IOMUXC_``` and ```imxX-iomux.h```. On the ```imx8m``` we don't have ```imx8m-iomux.h```, that's why I was okay with mixing ```I2C_CFG_``` and ```IOMUXC_```. Maybe I'm being a bit picky here, I don't know. I won't request any change for this on the new pull request.

**2/  Importing an imx6-iomux.h**
I asked you to **not** import anything from Code Aurora https://github.com/OP-TEE/optee_os/pull/4152#discussion_r511820100 but instead move existing ```IOMUXC_``` definitions from ```imx6.h``` to ```imx6-iomux.h```.
You noticed that many of definitions to be moved are unused : alright, let's remove them! https://github.com/OP-TEE/optee_os/pull/4152#discussion_r512504625 So we would end up with only i2c related definitions in  ```imx6-iomux.h```.
I will request the change on the new PR.

**3) dropping support for boot-loaders not setting the peripheral clock to 24MHz**
I provided the reason here : https://github.com/OP-TEE/optee_os/pull/4152#discussion_r512527245
Btw, could you answer my questions about this clock rate ? https://github.com/OP-TEE/optee_os/pull/4152#discussion_r512503493

See you on the other PR!

Comment (ldts) [reply core/include/drivers/imx_i2c.h:17]: I am getting so sick of github :) ...it is impossible to have a coherent conversation.

So to answer on the 24MHz, it is not something that I need for SE050 (we merged the patch I sent to u-boot to correct the clock issue so we are already ok in our baselines using the 24MHz clock). but I was thinking of other users that are booting with these old versions of u-boot....should not we be concerned about them?


btw u-boot are merging this patch so from the next release, the clock upstream will be 24MHz on imx6ull

**Pull request closed without merging**
